### PR TITLE
Patch aws-lambda instrumentation to support ESM

### PR DESCRIPTION
**Pull request by mxiamxia**
*Description of changes:*
1. Add `IS_ESM` check in `otel_instrument` wrapper before calling lambda handler.  If it is EMS format, add ESM node-options [supported](https://github.com/open-telemetry/opentelemetry-js/blob/966ac176af249d86de6cb10feac2306062846768/doc/esm-support.md#esm-options-for-different-versions-of-nodejs) by OTel community.  Added a new wrapper script `otel_instrument_esm` if the esm auto detection logic is failed, customer can opt-in to tell tell layer go with ESM instrumentation path.
2. Set a new env var `HANDLER_IS_ESM` for lambda function when ESM is detected
3. Patch aws-lambda instrumentation, when `IS_ESM` env var is set, apply ESM supported `InstrumentationNodeModuleDefinition` to patch function handler, otherwise keep using the existing handler patcher. 

Note: this change add a new branch for supporting ESM, the existing CommonJS path should not be impacted.

1. https://github.com/open-telemetry/opentelemetry-js/issues/4842
2. https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1942

*Test*
```

2024-10-09T20:38:13.411-07:00 | Instrumenting lambda handler {
-- | --
  | 2024-10-09T20:38:13.411-07:00 | taskRoot: '/var/task',
  | 2024-10-09T20:38:13.411-07:00 | handlerDef: 'index.handler',
  | 2024-10-09T20:38:13.411-07:00 | handler: 'index.handler',
  | 2024-10-09T20:38:13.411-07:00 | moduleRoot: '',
  | 2024-10-09T20:38:13.411-07:00 | module: 'index',
  | 2024-10-09T20:38:13.411-07:00 | filename: '/var/task/index.mjs',
  | 2024-10-09T20:38:13.411-07:00 | functionName: 'handler'
  | 2024-10-09T20:38:13.411-07:00 | }


  | 2024-10-09T20:38:15.386-07:00 | 'cloud.account.id': '252610625673',
-- | -- | --
  | 2024-10-09T20:38:15.386-07:00 | 'aws.is.local.root': true,
  | 2024-10-09T20:38:15.386-07:00 | 'aws.local.service': 'TestESM',
  | 2024-10-09T20:38:15.386-07:00 | 'aws.local.operation': 'TestESM/Handler',
  | 2024-10-09T20:38:15.386-07:00 | 'aws.span.kind': 'LOCAL_ROOT'


```
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lambda-layer/packages/layer/scripts/otel-instrument-esm [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+export NODE_OPTIONS="${NODE_OPTIONS} --import @aws/aws-distro-opentelemetry-node-autoinstrumentation/register --experimental-loader=@opentelemetry/instrumentation/hook.mjs"
+export LAMBDA_RESOURCE_ATTRIBUTES="cloud.region=$AWS_REGION,cloud.provider=aws,faas.name=$AWS_LAMBDA_FUNCTION_NAME,faas.version=$AWS_LAMBDA_FUNCTION_VERSION,faas.instance=$AWS_LAMBDA_LOG_STREAM_NAME,aws.log.group.names=$AWS_LAMBDA_LOG_GROUP_NAME";
+export IS_ESM=true
```
Comment (jj22ee) [inline lambda-layer/packages/layer/scripts/otel-instrument-esm:4]: Can we have more descriptive env var name, like `HANDLER_IS_ESM`?

File: lambda-layer/packages/layer/scripts/otel-instrument-esm [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+fi
+
+# - Set the service name
+if [ -z "${OTEL_SERVICE_NAME}" ]; then
+    export OTEL_SERVICE_NAME=$AWS_LAMBDA_FUNCTION_NAME;
+fi
+
+# - Set the propagators
+if [[ -z "$OTEL_PROPAGATORS" ]]; then
+  export OTEL_PROPAGATORS="tracecontext,baggage,xray"
```
Comment (jj22ee) [inline lambda-layer/packages/layer/scripts/otel-instrument-esm:23]: In `register.ts`, we use `xray,tracecontext,b3,b3multi`. Curious why there's a difference here?

File: lambda-layer/packages/layer/scripts/otel-instrument [hunk 3]
```diff
@@ -1,5 +1,40 @@
...
+
+  # Return true if both conditions are met
+  if $found_mjs || $is_module; then
+    return 0  # 0 in bash means true
+  else
+    return 1  # 1 in bash means false
+  fi
+}
+
+if isESMScript; then
```
Comment (jj22ee) [inline lambda-layer/packages/layer/scripts/otel-instrument:31]: Do we still need this, given we have a new `otel-instrument-esm` script? Or vice versa?

Comment (mxiamxia) [reply lambda-layer/packages/layer/scripts/otel-instrument:31]: yes. `otel-instrument-esm` is the fallback solution if the esm auto-detection doesn't work. We want to provide out of box CX on ESM support with the best efforts.

File: lambda-layer/packages/layer/scripts/otel-instrument-esm (hunk 2 above)
Comment (mxiamxia) [reply lambda-layer/packages/layer/scripts/otel-instrument-esm:23]: We're just following the existing OTel-lambda layer behavior on propagator settings for now
https://github.com/open-telemetry/opentelemetry-lambda/blob/51c8d3026742d1095f7f7614e88856c02dde49c4/nodejs/packages/layer/scripts/otel-handler#L12

**Review by jj22ee — APPROVED** [review]
TY! 🚢 
Great to have some ESM support for OTel in Lambda!

File: aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    const [module, functionName] = handler.split('.', 2);
+
+    // Lambda loads user function using an absolute path.
+    let filename = path.resolve(taskRoot, moduleRoot, module);
+    if (!filename.endsWith('.js')) {
+      // its impossible to know in advance if the user has a cjs or js or mjs file.
+      // check that the .js file exists otherwise fallback to next known possibility
+      try {
+        fs.statSync(`${filename}.js`);
```
Comment (pxaws) [inline aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts:42]: Does this require some permissions for nodejs to read the file? I guess this kind of permissions are already granted by lambda runtime?

File: aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        filename += '.js';
```
Comment (pxaws) [inline aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts:43]: I suggest to add some comments to indicate 
* ".js" can use both CJS and ESM
* ".cjs" must use CJS
* ".mjs" must use ESM

File: aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts (hunk 4 above)
Comment (mxiamxia) [reply aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/aws-lambda.ts:42]: yes. instrumentation and lambda function code are running on the same runtime and directory. it is mutual accessiable.

**Review by pxaws — APPROVED** [review]

**Pull request merged**